Scripts/Chunk.gd

```diff
@@ -28,28 +29,30 @@ func _generate_surface() -> Array:
 	var indices := PackedInt32Array()
 	var uvs := PackedVector2Array()
 	var triangle_index: int = 0
+	var mesh_simplification_increment: int = 1 if level_of_detail == 0 else level_of_detail * 2
+	var vertices_per_line: int = (chunk_data.size - 1) / mesh_simplification_increment + 1
 
-	vertices.resize(chunk_data.width * chunk_data.width)
-	indices.resize((chunk_data.width - 1) * (chunk_data.width - 1) * 6)
-	uvs.resize(chunk_data.width * chunk_data.width)
+	vertices.resize(vertices_per_line * vertices_per_line)
+	indices.resize((vertices_per_line - 1) * (vertices_per_line - 1) * 6)
+	uvs.resize(vertices_per_line * vertices_per_line)
 
 	surface_array.resize(ArrayMesh.ARRAY_MAX)
 
-	for z in range(chunk_data.width):
-		for x in range(chunk_data.width):
-			var index: int = x + z * chunk_data.width
+	for z in range(0, chunk_data.size, mesh_simplification_increment):
+		for x in range(0, chunk_data.size, mesh_simplification_increment):
+			var index: int = (x / mesh_simplification_increment) + (z / mesh_simplification_increment) * vertices_per_line
 			var height: float = chunk_data.height_map[index]
```

Comment:
This is incorrect, as you height_map is still full sized.
You need to either update how you access the index of the height_map:
```
var height_index: int = x + z * chunk_data.width
```

Or have the LOD be considered to update the height_map array resolution to match the vertices.